### PR TITLE
Add set custom header to Seedling Runtime

### PR DESCRIPTION
**Pull request by NachoPal**
Please review and ONLY MERGE AFTER: #845 

These changes are not easy to test and I didn't find time for it yet.

Let me know if this is the correct approach and if it should work

### CODE REVIEW

File: pallets/parachain-system/src/lib.rs [hunk 1]
```diff
@@ -328,6 +328,15 @@ pub mod pallet {
...
 
 					Self::put_parachain_code(&validation_code);
+
+					if let Some(head_data) = <PendingCustomValidationHeadData<T>>::get() {
+						Self::set_custom_validation_head_data(head_data);
+						<PendingCustomValidationHeadData<T>>::kill();
+						Self::deposit_event(Event::CustomValidationHeadDataApplied(
+							vfp.relay_parent_number,
+						));
+					}
```
Comment (bkchr) [inline pallets/parachain-system/src/lib.rs:338]: This should be moved out to the other pallet.

File: pallets/parachain-system/src/lib.rs [hunk 2]
```diff
@@ -935,6 +953,13 @@ impl<T: Config> Pallet<T> {
 	pub fn set_custom_validation_head_data(head_data: Vec<u8>) {
 		CustomValidationHeadData::<T>::put(head_data);
 	}
+
+	/// Set a custom head data that should only be applied when upgradeGoAheadSignal from
+	/// the Relay Chain is GoAhead
+	pub fn set_pending_custom_validation_head_data(head_data: Vec<u8>) {
+		PendingCustomValidationHeadData::<T>::put(head_data);
+		Self::deposit_event(Event::CustomValidationHeadDataStored);
+	}
```
Comment (bkchr) [inline pallets/parachain-system/src/lib.rs:962]: Same

File: pallets/solo-to-para/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (bkchr) [inline pallets/solo-to-para/src/lib.rs:1]: License header missing.

File: pallets/solo-to-para/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Set a custom head data that should only be applied when upgradeGoAheadSignal from
+		/// the Relay Chain is GoAhead
+		fn store_pending_custom_validation_head_data(head_data: Vec<u8>) {
+			PendingCustomValidationHeadData::<T>::put(head_data);
+			Self::deposit_event(Event::CustomValidationHeadDataStored);
```
Comment (bkchr) [inline pallets/solo-to-para/src/lib.rs:92]: Second event that is being send as part of `schedule_migration`

File: pallets/solo-to-para/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+		StorageValue<_, Vec<u8>, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event {
+		MigrationScheduled,
+		/// The custom validation head data has been scheduled to apply.
+		CustomValidationHeadDataStored,
+		/// The custom validation head data was applied as of the contained relay chain block number.
+		CustomValidationHeadDataApplied, //(RelayChainBlockNumber),
```
Comment (bkchr) [inline pallets/solo-to-para/src/lib.rs:60]:
```suggestion
		CustomValidationHeadDataApplied,
```

File: pallets/solo-to-para/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+	impl<T: Config> Pallet<T> {
+		/// Set a custom head data that should only be applied when upgradeGoAheadSignal from
+		/// the Relay Chain is GoAhead
+		fn store_pending_custom_validation_head_data(head_data: Vec<u8>) {
+			PendingCustomValidationHeadData::<T>::put(head_data);
+			Self::deposit_event(Event::CustomValidationHeadDataStored);
+		}
+
+		pub fn set_pending_custom_validation_head_data() -> Result<(), Error<T>> {
```
[the remark this answers is not in the document]
Comment (riusricardo) [reply pallets/solo-to-para/src/lib.rs:95]: Somewhere around here: https://github.com/paritytech/cumulus/blob/3ba4446377d44cbb7ba41f62fda0050adae9ea28/pallets/parachain-system/src/lib.rs#L322-L333

File: pallets/parachain-system/src/lib.rs [hunk 7]
```diff
@@ -882,7 +882,7 @@ impl<T: Config> Pallet<T> {
 	}
 
 	/// The implementation of the runtime upgrade functionality for parachains.
-	fn set_code_impl(validation_function: Vec<u8>) -> DispatchResult {
+	pub fn set_code_impl(validation_function: Vec<u8>) -> DispatchResult {
```
[the remark this answers is not in the document]
Comment (NachoPal) [reply pallets/parachain-system/src/lib.rs]: I changed it to `schedule_code_upgrade`. Just notice that if this is merged we have to modify also the `core-contributior-docs/docs/learn-cumulus/cumulus-overview.md`: https://github.com/search?q=org%3Aparitytech+set_code_impl&type=code

File: pallets/solo-to-para/src/lib.rs (hunk 4 above)
Comment (NachoPal) [reply pallets/solo-to-para/src/lib.rs:90]: I removed one of the events. Now we emit only `CustomValidationHeadDataStored`

File: pallets/parachain-system/src/lib.rs [hunk 8]
```diff
@@ -328,6 +328,7 @@ pub mod pallet {
 					let validation_code = <PendingValidationCode<T>>::take();
 
 					Self::put_parachain_code(&validation_code);
+					<T::OnValidationData as OnValidationData>::on_validation_code_applied();
```
Comment (bkchr) [inline pallets/parachain-system/src/lib.rs:331]: Trait and associated type should be renamed.

File: primitives/core/src/lib.rs [hunk 9]
```diff
@@ -142,6 +142,7 @@ pub enum ServiceQuality {
 #[impl_trait_for_tuples::impl_for_tuples(30)]
 pub trait OnValidationData {
 	fn on_validation_data(data: &PersistedValidationData);
+	fn on_validation_code_applied();
```
Comment (bkchr) [inline primitives/core/src/lib.rs:145]: Some docs would be nice. (I know that the old function also doesn't has any docs :see_no_evil: )

File: pallets/solo-to-para/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+			if let Some(head_data) = <PendingCustomValidationHeadData<T>>::get() {
+				parachain_system::Pallet::<T>::set_custom_validation_head_data(head_data);
+				<PendingCustomValidationHeadData<T>>::kill();
+				Self::deposit_event(Event::CustomValidationHeadDataApplied);
+			}
+		}
+	}
+
+	pub struct ValidationData<T: Config>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnValidationData for ValidationData<T> {
```
Comment (bkchr) [inline pallets/solo-to-para/src/lib.rs:106]: You can just implement the trait for the pallet.

File: pallets/parachain-system/src/lib.rs (hunk 8 above)
Comment (NachoPal) [reply pallets/parachain-system/src/lib.rs:332]: Sorry, I don't get exactly how to change this one. Can you please propose the change to be merged?

File: pallets/solo-to-para/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,183 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline pallets/solo-to-para/src/lib.rs:1]:
```suggestion
// Copyright 2022 Parity Technologies (UK) Ltd.
```

File: primitives/core/src/lib.rs [hunk 12]
```diff
@@ -142,7 +142,10 @@ pub enum ServiceQuality {
 /// A trait which is called when the validation data is set.
 #[impl_trait_for_tuples::impl_for_tuples(30)]
 pub trait OnValidationData {
```
Comment (bkchr) [inline primitives/core/src/lib.rs:144]: We should change the name. I don't have any good idea, but it is clearly not only about `OnValidationData` anymore.

File: pallets/solo-to-para/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+	impl<T: Config> Pallet<T> {
+		/// Set a custom head data that should only be applied when upgradeGoAheadSignal from
+		/// the Relay Chain is GoAhead
+		fn store_pending_custom_validation_head_data(head_data: Vec<u8>) {
+			PendingCustomValidationHeadData::<T>::put(head_data);
+			Self::deposit_event(Event::CustomValidationHeadDataStored);
+		}
+
+		/// Apply the peding custom head data that the Relay Chain should expect next
+		pub fn set_pending_custom_validation_head_data() {
```
Comment (bkchr) [inline pallets/solo-to-para/src/lib.rs:96]:
```suggestion
		/// Set pending custom head data as head data that will be returned by `validate_block`. on the relay chain.
		fn set_pending_custom_validation_head_data() {
```

File: pallets/solo-to-para/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+			if let Some(head_data) = <PendingCustomValidationHeadData<T>>::get() {
+				parachain_system::Pallet::<T>::set_custom_validation_head_data(head_data);
+				<PendingCustomValidationHeadData<T>>::kill();
+				Self::deposit_event(Event::CustomValidationHeadDataApplied);
+			}
```
Comment (bkchr) [inline pallets/solo-to-para/src/lib.rs:101]:
```suggestion
			if let Some(head_data) = <PendingCustomValidationHeadData<T>>::take() {
				parachain_system::Pallet::<T>::set_custom_validation_head_data(head_data);
				Self::deposit_event(Event::CustomValidationHeadDataApplied);
			}
```

File: pallets/parachain-system/src/lib.rs (hunk 8 above)
Comment (bkchr) [reply pallets/parachain-system/src/lib.rs:332]: The associoated type `type OnValidationData` should be renamed, similar to the trait as I have said below.

File: primitives/core/src/lib.rs (hunk 12 above)
Comment (NachoPal) [reply primitives/core/src/lib.rs:144]: I don't have ideas either... It is still called inside the `set_validation_data` function... so for me it still makes sense to keep the name, or at least is not so bad.

Comment (bkchr) [reply primitives/core/src/lib.rs:144]:
```
/// Something that should be informed about system related events.
///
/// This includes events like [`on_validation_data`](Self::on_validation_data) that is being called when
/// the parachain inherent is executed that contains the validation data. Or like [`on_validation_code_applied`](Self::on_validation_code_applied) that is called when the new validation is written to the state. This means that
/// from the next block the runtime is being using this new code.
#[impl_trait_for_tuples::impl_for_tuples(30)]
trait OnSystemEvent {
	/// Called in each blocks once when the validation data is set by the inherent.
	fn on_validation_data(data: &PersistedValidationData) {}
	/// Called when the validation code is being applied, aka from the next block on this is the new runtime.
	fn on_validation_code_applied()  {}
}
```

Please move the trait to `parachain-system`. Please also rename the associated type to `OnSystemEvent`

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
@NachoPal feel free to merge as you like.

**Pull request merged**